flake.nix: rewrite

philiptaron opened the pull request:
@SomeoneSerge and I rewrote `flake.nix` with the changes listed below. The top line from a user perspective is:

* Building CUDA-accelerated packages now works (`nix run github:ggerganov/llama.cpp#cuda`, once this PR is merged) without asking to accept unfree packages.
* There's support for [NVIDIA Jetson](https://www.nvidia.com/en-us/autonomous-machines/embedded-systems).
* When using ROCm and CUDA acceleration, _all_ packages that might are in the dependency graph are built with that acceleration, most importantly openmpi. 
* For platforms like MacOS where the CUDA or ROCm backend isn't supported, it's not listed.
* There are extra dev shells, one per backend.
* For users integrating this flake into a NixOS system, there's an overlay exposing the `llama-cpp` package and `llamaPackages` scope.

## Detailed change log

1. Split into separate files per output, using the pattern established at [flake.parts](https://flake.parts). 
2. Added an overlay, so that this flake can be integrated into others. 
3. In the overlay, added a scope called `llamaPackages` which can contain all the relevant packages.
4. The name in the overlay is `llama-cpp`. The packages `opencl`, `cuda`, and `rocm` are derived from this one, so that they fit into the broader set of Nix packages from [nixpkgs](https://github.com/nixos/nixpkgs).
5. Use [callPackage](https://summer.nixos.org/blog/callpackage-a-tool-for-the-lazy/) rather than `with pkgs;` so that there's dependency injection rather than dependency lookup.
6. Add a description and meta information for each package. The description includes a bit about what's trying to accelerate each one. @SomeoneSerge and I are listed as maintainers, which means we're mildly on the hook if something about this flake breaks.
7. The flake is published to https://flakestry.dev/.
8. The size of the dependency graph (the build and runtime closure) is down about 1.4 GB due to being more precise about dependencies (we now use specific CUDA packages instead of cudatoolkit).
9. We now use the apple_sdk selection logic in nixpkgs instead of rolling our own.
10. Accept the CUDA packages for `nix run` use. When used through the overlay, they still need to be accepted by the greater system, obviously.
11. Updated flake.lock to pull in the latest nixpkgs.
12. Formatted the Nix code with the in-progress [RFC101](https://github.com/nix-rfc-101/rfcs/blob/master/rfcs/0166-nix-formatting.md) formatter. `nix shell github:piegamesde/nixfmt/rfc101-style --command nixfmt flake.nix .devops/nix/*.nix`

I tested this out using [DaringMaid-13B-GGUF](https://huggingface.co/TheBloke/DaringMaid-13B-GGUF) on NixOS x86_64-linux with all the variants, including CUDA, and macOS Sonoma aarch64-darwin. While the flake claims to support macOS x86_64-darwin and aarch64-linux, I don't have a machine to support building and running these variants.

To run it yourself, [install Nix](https://github.com/DeterminateSystems/nix-installer) and run `nix run github:philiptaron/llama.cpp/nix`. Use `nix flake show github:philiptaron/llama.cpp/nix` to see what the flake has in it.

philiptaron committed:
flake.nix: rewrite

1. Split into separate files per output.

2. Added overlays, so that this flake can be integrated into others.
   The names in the overlay are `llama-cpp`, `llama-cpp-opencl`,
   `llama-cpp-cuda`, and `llama-cpp-rocm` so that they fit into the
   broader set of Nix packages from [nixpkgs](https://github.com/nixos/nixpkgs).

3. Use [callPackage](https://summer.nixos.org/blog/callpackage-a-tool-for-the-lazy/)
   rather than `with pkgs;` so that there's dependency injection rather
   than dependency lookup.

4. Add a description and meta information for each package.
   The description includes a bit about what's trying to accelerate each one.

5. Use specific CUDA packages instead of cudatoolkit on the advice of SomeoneSerge.

6. Format with `serokell/nixfmt` for a consistent style.

7. Update `flake.lock` with the latest goods.

philiptaron commented:
@Green-Sky , you might be interested.

ggerganov commented:
Is it possible to move some (or all) of these files into a sub-folder? Maybe in `.devops` even. Want to keep the root less cluttered. If Nix does not allow it, then I guess it's fine like that.

Will wait for a review from somebody as I am not familiar with Nix

Green-Sky commented:
Nice, someone who knows how to properly write some nix. :) This is a very nice pr.

> Is it possible to move some (or all) of these files into a sub-folder? Maybe in .devops even. Want to keep the root less cluttered.

Yes, the `flake.nix` is just the entry point, and it imports the other files, so putting the other files into eg. `.devops/nix/` would be nice.

@philiptaron Is there a reason no `flake.lock` file is checked in? I like to keep the same lock file around for a while, since the cuda framework is a pain to install. It is not cached (non-free) + it takes 20min to download and install everytime + its not very small and takes up alot of space.

Green-Sky commented:
I hit an error with cuda. It looks like we need to restrict it to gcc11:
```
-- Found CUDAToolkit: /nix/store/dxwbj894wafwqq2dmcxsr9sb6sf546p7-cudatoolkit-11.8.0-merged/include (found version "11.8.89")
-- cuBLAS found
CMake Error at /nix/store/4vq5ggsg1vmfs09r4sqbidmgvqlxrv14-cmake-3.27.8/share/cmake-3.27/Modules/CMakeDetermineCompilerId.cmake:753 (message):
  Compiling the CUDA compiler identification source file
  "CMakeCUDACompilerId.cu" failed.

  Compiler:
  /nix/store/dxwbj894wafwqq2dmcxsr9sb6sf546p7-cudatoolkit-11.8.0-merged/bin/nvcc
```
...
```
  /nix/store/gg7gma5n3q9xdqgxns3d1wz0xn857m04-cudatoolkit-11.8.0/include/crt/host_config.h:132:2:
  error: #error -- unsupported GNU version! gcc versions later than 11 are
  not supported! The nvcc flag '-allow-unsupported-compiler' can be used to
  override this version check; however, using an unsupported host compiler
  may cause compilation failure or incorrect run time execution.  Use at your
  own risk.

    132 | #error -- unsupported GNU version! gcc versions later than 11 are not supported! The nvcc flag '-allow-unsupported-compiler' can be used to override this version c>
        |  ^~~~~

  # --error 0x1 --
```

looks like we need to use the `stdenv` from https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/cuda-modules/backend-stdenv.nix or something.

SomeoneSerge commented:
> looks like we need to use the stdenv from https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/cuda-modules/backend-stdenv.nix or something.

Yes, the current solution is to do something in the following lines:

https://github.com/NixOS/nixpkgs/blob/e3f251c662bc525e4bae3edfa3fc67e52d690d4f/pkgs/development/libraries/opencv/4.x.nix#L90-L93

We're not documenting this, because we're hoping to make it unnecessary to do this soon-ish

SomeoneSerge commented on overlays.nix:
Nit: `overlays` is plural, but the content is a single overlay

SomeoneSerge commented on llama-cpp.nix:
```diff
@@ -0,0 +1,118 @@
+, CoreVideo ? null
+, CoreGraphics ? null
+, useOpenCL ? false
+, useCuda ? false
+, useRocm ? false
```
`useCuda ? config.cudaSupport`, `useRocm ? config.rocmSupport`

SomeoneSerge commented on llama-cpp.nix:
```diff
@@ -0,0 +1,118 @@
+  isDefault = !useOpenCL && !useCuda && !useRocm;
+
+  # Give a little description difference between the flavors.
+  descriptionSuffix = if useOpenCL then
+    " (OpenCL accelerated)"
+  else if useCuda then
+    " (CUDA accelerated)"
+  else if useRocm then
+    " (ROCm accelerated)"
+  else if (MetalKit != null) then
+    " (MetalKit accelerated)"
+  else "";
```
Nit: `optionalString useXXX "..." + optionalString useYYY "..."` might look cleaner in the end

philiptaron replied:
I decided after trying it out to keep what I originally had. Another PR can change it.

SomeoneSerge commented on llama-cpp.nix:
```diff
@@ -0,0 +1,118 @@
+, ninja
+, pkg-config
+, symlinkJoin
+, python3
```
When packaging a `buildPythonApplication`, it's preferable to take `python3Packages` as argument (you can access the interpreter as `python3Packages.python`) because of splicing

philiptaron replied:
I think I'm going to leave this one as-is from the previous flake.nix, since I don't _really_ understand what better looks like. We can always continue to iterate on it.

SomeoneSerge replied:
Ah sorry, missed your comment. Sure, we can change that later. I linked the documentation in the other message though

SomeoneSerge commented on llama-cpp.nix:
```diff
@@ -0,0 +1,118 @@
+  rocmBuildInputs = with rocmPackages; [ clr hipblas rocblas ];
+in
+
+stdenv.mkDerivation {
```
Maybe out of scope of this PR, but it'd be nice to migrate to `buildPythonPackage`, as well as ensure this repo is pep517-compatible, and ensure the correct `.dist-info` is generated

SomeoneSerge replied:
Ah, nevermind, I see now that the python bindings are in a separate project

SomeoneSerge commented on llama-cpp.nix:
```diff
@@ -0,0 +1,118 @@
+  postPatch = ''
+    substituteInPlace ./ggml-metal.m \
+      --replace '[bundle pathForResource:@"ggml-metal" ofType:@"metal"];' "@\"$out/bin/ggml-metal.metal\";"
+    substituteInPlace ./*.py --replace "/usr/bin/env python" "${llama-python}/bin/python"
```
If we were to migrate to `buildPythonPackage` and prepare the pyproject.toml we could make those into setuptools' entrypoints

philiptaron replied:
Good idea; I'll leave it for future PRs.

SomeoneSerge commented on llama-cpp.nix:
```diff
@@ -0,0 +1,118 @@
+
+  postPatch = ''
+    substituteInPlace ./ggml-metal.m \
+      --replace '[bundle pathForResource:@"ggml-metal" ofType:@"metal"];' "@\"$out/bin/ggml-metal.metal\";"
```
I'm not familiar with this, is `[bundle pathForResource:@"default" ofType:@"metallib"];` some standard syntax? Is it only with the nix package that we have to override this, or all users have to do that? In the latter case, could we address this at the source level? E.g. make it so this could be specified in the cmake options (`cmakeFlags`)?

philiptaron replied:
I'm also not familiar with what this is doing. I ported it over wholesale from @Green-Sky's original `flake.nix`.

SomeoneSerge commented on llama-cpp.nix:
```diff
@@ -0,0 +1,118 @@
+  ++ optionals isDefault defaultBuildInputs;
+
+  cmakeFlags = [
+    "-DLLAMA_NATIVE=OFF"
+    "-DLLAMA_BUILD_SERVER=ON"
+    "-DBUILD_SHARED_LIBS=ON"
+    "-DCMAKE_SKIP_BUILD_RPATH=ON"
```
Nit: `lib.cmakeBool` and `lib.cmakeFeature`

SomeoneSerge commented on llama-cpp.nix:
```diff
@@ -0,0 +1,118 @@
+    # Build all targets supported by rocBLAS. When updating search for TARGET_LIST_ROCM
+    # in https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/develop/CMakeLists.txt
+    # and select the line that matches the current nixpkgs version of rocBLAS.
+    "-DAMDGPU_TARGETS=gfx803;gfx900;gfx906:xnack-;gfx908:xnack-;gfx90a:xnack+;gfx90a:xnack-;gfx940;gfx941;gfx942;gfx1010;gfx1012;gfx1030;gfx1100;gfx1101;gfx1102"
```
Nit: make it a list and an argument to `callPackage`. Looking at the expression for pytorch in Nixpkgs, you could take `rocmPackages.clr.gpuTargets` as the default value.

SomeoneSerge replied:
On this note, if you want this package to respect `config.cudaCapabilities`, you currently have to specify those manually as well (we might automate this for FindCUDAToolkit.cmake later)

philiptaron replied:
```
$ nix eval nixpkgs#rocmPackages.clr.gpuTargets
[ "gfx803" "gfx900" "gfx906" "gfx908" "gfx90a" "gfx940" "gfx941" "gfx942" "gfx1010" "gfx1012" "gfx1030" "gfx1100" "gfx1101" "gfx1102" ]
```

It's missing the `xnack-` entries, but the other entries are the same. I think I'll leave this to someone who actually owns a Radeon card and knows what's going on with ROCm, though I agree that this looks super gross right now.

Madouura replied:
I suggest just using ``rocmPackages.clr.gpuTargets`` and appending the xnack variants.

SomeoneSerge commented on overlays.nix:
```diff
@@ -0,0 +1,20 @@
+let
+  inherit (prev.stdenv) isAarch32 isAarch64 isDarwin;
+
+  darwinSpecific = if isAarch64 then {
+    inherit (prev.darwin.apple_sdk_11_0.frameworks) Accelerate MetalKit;
+  } else {
+    inherit (prev.darwin.apple_sdk.frameworks) Accelerate CoreGraphics CoreVideo;
+  };
```
Nit: could do

```
apple_sdk = if ... then prev.darwin.apple_sdk_11_0 else prev.darwin_apple_sdk;
inherit (apple_sdk) ...;
```

(on that note, I'm not sure why do we pass these as individual arguments instead of passing `apple_sdk` as the whole thing; outside my expertise though)

philiptaron replied:
Mine too. I'm keeping it from the previous impl, but I don't really [understand the structure](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/darwin/apple-sdk/frameworks.nix#L2).

SomeoneSerge commented on overlays.nix:
```diff
@@ -0,0 +1,20 @@
+in
+
+{
+  llama-cpp = prev.callPackage ./llama-cpp.nix osSpecific;
+  llama-cpp-opencl = prev.callPackage ./llama-cpp.nix (osSpecific // { useOpenCL = true; });
+  llama-cpp-cuda = prev.callPackage ./llama-cpp.nix (osSpecific // { useCuda = true; });
+  llama-cpp-rocm = prev.callPackage ./llama-cpp.nix (osSpecific // { useRocm = true; });
```
An alternative to `osSpecific // ...` could be to define `llama-cpp = callPackage ...` once and then use `llama-cpp-xxx = final.llama-cpp.override ...`. This has pros and cons, e.g. the users can apply an overlay to `llama-cpp` once and have the effect propagate to all the variants at once. The cost is that it's easier to get into an infinite recursion I guess

philiptaron replied:
That con is why I picked this structure. We can always refactor it into a different shape if the pros weigh out.

SomeoneSerge commented on llama-cpp.nix:
```diff
@@ -0,0 +1,118 @@
+    CoreGraphics
+  ];
+
+  cudaBuildInput = symlinkJoin (with cudaPackages; {
```
IIRC llama.cpp uses `find_package(CUDAToolkit)`. This means you don't have to use `symlinkJoin` nor `cudaPackages.cudatoolkit`. Both are expensive and discouraged (mostly because of the huge runtime closure).

Cf. e.g. pytorch from Nixpkgs:

https://github.com/NixOS/nixpkgs/blob/e3f251c662bc525e4bae3edfa3fc67e52d690d4f/pkgs/development/libraries/opencv/4.x.nix#L360-L361
https://github.com/NixOS/nixpkgs/blob/e3f251c662bc525e4bae3edfa3fc67e52d690d4f/pkgs/development/libraries/opencv/4.x.nix#L365-L367
https://github.com/NixOS/nixpkgs/blob/e3f251c662bc525e4bae3edfa3fc67e52d690d4f/pkgs/development/libraries/opencv/4.x.nix#L303
https://github.com/NixOS/nixpkgs/blob/e3f251c662bc525e4bae3edfa3fc67e52d690d4f/pkgs/development/libraries/opencv/4.x.nix#L333-L356

SomeoneSerge replied:
Oh, I was hoping github would render the previews

philiptaron replied:
Great. I took the suggestion and added the specific build and runtime dependencies needed. It's much cleaner now.

SomeoneSerge commented on llama-cpp.nix:
```diff
@@ -0,0 +1,118 @@
+    # and select the line that matches the current nixpkgs version of rocBLAS.
+    "-DAMDGPU_TARGETS=gfx803;gfx900;gfx906:xnack-;gfx908:xnack-;gfx90a:xnack+;gfx90a:xnack-;gfx940;gfx941;gfx942;gfx1010;gfx1012;gfx1030;gfx1100;gfx1101;gfx1102"
+  ]
+  ++ optionals isDefault (if (MetalKit != null) then [
+    "-DCMAKE_C_FLAGS=-D__ARM_FEATURE_DOTPROD=1"
+    "-DLLAMA_METAL=ON"
+  ] else [
+    "-DLLAMA_BLAS=ON"
+    "-DLLAMA_BLAS_VENDOR=OpenBLAS"
+  ]);
```
A possible alternative:

```
optionals isDefault [
  (lib.cmakeBool "LLAMA_METAL" (MetalKit != null))
  (lib.cmakeBool "LLAMA_BLAS" (MetalKit == null))
  # ...
]
```

philiptaron replied:
Nice. It's cleaner.

SomeoneSerge commented on llama-cpp.nix:
```diff
@@ -0,0 +1,118 @@
+  ]);
+
+  postInstall = ''
+    mv $out/bin/main $out/bin/llama
+    mv $out/bin/server $out/bin/llama-server
+    mkdir -p $out/include
+    cp $src/llama.h $out/include/
```
It's better to add proper install targets at the cmake level, if they haven't been added yet. An example (a bit unorderly) https://github.com/SomeoneSerge/pkgs/blob/628b080165cde1528c2d40249cef114cf2cec3e2/pkgs/by-name/ll/llama-cpp/package.nix#L42-L56

philiptaron replied:
I'll leave that for another PR by someone more experienced in CMake than I. 🧑‍🎓

SomeoneSerge commented:
> Is it possible to move some (or all) of these files into a sub-folder? 

I think we should. We'd want to keep `flake.nix` and `flake.lock` at the top-level though, is that alright?

By the way, @ggerganov how would you feel about adding a github workflow for `nix-build` to have these automatically checked? I think these shouldn't be expensive, especially if we also enable https://github.com/DeterminateSystems/magic-nix-cache-action and https://github.com/marketplace/actions/cachix.

philiptaron commented on flake.lock:
https://github.com/ggerganov/llama.cpp/pull/4605#issuecomment-1868267460 @Green-Sky here's the `flake.lock`.

Green-Sky replied:
Oh, the github diff looked like the file was removed at a glance. Thanks for cleaning it.

philiptaron commented:
> Is it possible to move some (or all) of these files into a sub-folder? Maybe in `.devops` even.

Done. Thanks for the suggestion.

Green-Sky commented:
> how would you feel about adding a github workflow for nix-build to have these automatically checked? I think these shouldn't be expensive, especially if we also enable https://github.com/DeterminateSystems/magic-nix-cache-action and https://github.com/marketplace/actions/cachix.

That is a good idea, microsoft is paying the bill for this anyway.

ParetoOptimalDev commented:
> > how would you feel about adding a github workflow for nix-build to have these automatically checked? I think these shouldn't be expensive, especially if we also enable https://github.com/DeterminateSystems/magic-nix-cache-action and https://github.com/marketplace/actions/cachix.
> 
> That is a good idea, microsoft is paying the bill for this anyway.

If you want things to be as incremental as possible, see:

https://github.com/Mic92/nix-fast-build

For instance see this quote about the effectiveness of it's caching:

> So evaluating [disko](https://github.com/nix-community/disko), nix-ci-build takes for an already build flake 10.929s and nixci uses 1:54.75 minutes for the same flake. This is because it can build and evaluate in parallel (building already starts when the flake is still evaluate). I know that [devour-flake](https://github.com/srid/devour-flake) is meant to provide caching but if you change a single byte, which is what I do in my project when testing changes, than this cache is no longer effective.

ParetoOptimalDev commented:
This rewrite and the current flake aren't meant to run the openai compatible server are they?

https://github.com/ggerganov/llama.cpp/blob/master/examples/server/api_like_OAI.py

philiptaron commented:
> This rewrite and the current flake aren't meant to run the openai compatible server are they?
> 
> 
> 
> https://github.com/ggerganov/llama.cpp/blob/master/examples/server/api_like_OAI.py

My main goal is to enable integration into a NixOS system, and to make the build process for CUDA in particular more straightforward.

That should enable future PRs to package the Python bits in this repository... but doesn't do that packaging at this point.

ParetoOptimalDev commented:
> > This rewrite and the current flake aren't meant to run the openai compatible server are they?
> > https://github.com/ggerganov/llama.cpp/blob/master/examples/server/api_like_OAI.py
> 
> My main goal is to enable integration into a NixOS system, and to make the build process for CUDA in particular more straightforward.
> 
> That should enable future PRs to package the Python bits in this repository... but doesn't do that packaging at this point.

I submitted this because I needed it now and didn't want the functionality lost to the sands of time:

https://github.com/ggerganov/llama.cpp/pull/4612

However I'm sure it'll be easier after your rewrite PR here.

SomeoneSerge commented on .devops/overlay.nix:
```diff
@@ -0,0 +1,20 @@
+in
+
+{
+  llama-cpp = prev.callPackage ./package.nix osSpecific;
+  llama-cpp-opencl = prev.callPackage ./package.nix (osSpecific // { useOpenCL = true; });
+  llama-cpp-cuda = prev.callPackage ./package.nix (osSpecific // { useCuda = true; });
+  llama-cpp-rocm = prev.callPackage ./package.nix (osSpecific // { useRocm = true; });
```
You almost never want to use `prev.callPackage` over `final.callPackage`. The heuristic is that you want the final versions of your dependencies.

And just in case that got lost, there's an alternative to multiple `callPackage`s, which is to use `.override`. Example: https://github.com/NixOS/nixpkgs/blob/0fe31382c0756f500b0cc6ddc7c1965afd28a983/pkgs/top-level/python-packages.nix#L14350-L14364.

philiptaron replied:
Thanks, this was really helpful. Updated.

SomeoneSerge commented on .devops/overlay.nix:
```diff
@@ -0,0 +1,20 @@
+  darwinSpecific = if isAarch64 then {
+    inherit (prev.darwin.apple_sdk_11_0.frameworks) Accelerate MetalKit;
+  } else {
+    inherit (prev.darwin.apple_sdk.frameworks) Accelerate CoreGraphics CoreVideo;
```
Same comment about `prev` v. `final`

SomeoneSerge commented on .devops/package.nix:
```diff
@@ -0,0 +1,119 @@
+, ninja
+, pkg-config
+, git
+, python3
```
https://github.com/NixOS/nixpkgs/commit/70aa345bec1135ea9524cd893aea1f6e3dd82387

SomeoneSerge commented on .devops/package.nix:
```diff
@@ -0,0 +1,119 @@
+  '';
+
+  nativeBuildInputs = [ cmake ninja pkg-config git ]
+  ++ optional useCuda [ cudaPackages.cuda_nvcc ];
```
`optionals :: Bool -> [a] -> [a]`
`optional :: Bool -> a -> [a]`

SomeoneSerge commented on .devops/package.nix:
```diff
@@ -0,0 +1,119 @@
+    "-DLLAMA_BLAS_VENDOR=OpenBLAS"
+  ]);
+
+  postInstall = ''
+    mv $out/bin/main $out/bin/llama
+    mv $out/bin/server $out/bin/llama-server
+    mkdir -p $out/include
+    cp $src/llama.h $out/include/
+  '';
```
You're right about leaving the install targets for future work. Could you still add a comment about that just so the next person knows where our thinking was headed?

SomeoneSerge commented:
@philiptaron thank you for all the changes. There are many things to consider for future work, but I think this PR looks mostly finished now! I'm especially happy you got this to work with the splayed cuda, this saves a lot of storage and traffic 🙃.

I'd maybe consider nesting the nix files even further like in `nix/` or `.devops/nix` so they don't mix visually with the other CI stuff. Having to use the `../`-relative paths is sort of un-nice but certainly  affordable

SomeoneSerge commented on .devops/package.nix:
```diff
@@ -0,0 +1,119 @@
+, pkg-config
+, git
+, python3
+, openmpi
```
Llama.cpp seems to use implementation-agnostic apis, so we don't have to hardcode openmpi (there's a top-level `mpi` attribute which the users are free to reset to e.g. mpich instead of openmpi): https://github.com/ggerganov/llama.cpp/blob/708e179e8562c2604240df95a2241dea17fd808b/CMakeLists.txt#L324-L326

SomeoneSerge commented on .devops/package.nix:
```diff
@@ -0,0 +1,119 @@
+, git
+, python3
+, openmpi
+, openblas
```
Nixpkgs also provides the "shim" attributes `blas` and `lapack` which generate implementation agnostic pkg-config files, but let the user switch between openblas, mkl, etc.

This one I'm not sure will work out of the box, but you could leave a comment about trying the generic `blas` instead of `openblas` in future

philiptaron replied:
You were right. I added a comment.

SomeoneSerge commented on flake.nix:
```diff
@@ -1,139 +1,55 @@
+
+    overlay = import ./.devops/overlay.nix;
+
+    importNixpkgs = system: import nixpkgs {
+      inherit system;
+      overlays = [ overlay ];
+    };
```
H'm. Note that in the current form `github:ggerganov/llama.cpp#cuda` remains "broken" because the licenses have to be accepted at the nixpkgs import/instantiation time. I.e. you'd have to reimport nixpkgs with `config.allowUnfree = true`* for `packages.${system}.cuda` to make it work "out of the box" and in the pure evaluation mode

`*`: if you wanted to keep it precise, you could instead use something like `config.allowUnfreePredicate = p: builtins.all (license: license.free || builtins.elem license.shortName [ "CUDA EULA" "cuDNN EULA" ]) (p.meta.licenses or [p.meta.license])`.

I'd also import nixpkgs with `config.cudaSupport = true` to ensure that the dependencies (e.g. openmpi depends on `ucx` and `ucc`) use CUDA consistently as well.

I'd rather not postpone these two until later PRs.

In future we could also consider issues like those described in https://zimbatm.com/notes/1000-instances-of-nixpkgs, but right now that's strictly out of scope

philiptaron replied:
I rewrote this section of the flake (and removed the overlays that were pre-customized.)  It both appears to build and to test.

philiptaron commented:
OK. I've incorporated the feedback from the last round; thank you @SomeoneSerge for the kind and detailed look through. It's ready for a third look.

ggerganov reviewed (approved):
Adding additional CI workflows is totally fine 👍

SomeoneSerge committed:
flake.nix: use finalPackage instead of passing it manually

SomeoneSerge committed:
nix: unclutter darwin support

SomeoneSerge committed:
nix: pass most darwin frameworks unconditionally

...for simplicity

philiptaron commented:
I just tested these latest changes with my M2 Mac and my NixOS CUDA box using https://huggingface.co/TheBloke/DaringMaid-13B-GGUF. Everything appears to work swimmingly.

SomeoneSerge reviewed (approved):
LGTM, thanks @philiptaron! I'd squash the `fixup!` commit, but all the others "we could also do" we can do in later PRs

```
❯ nix build github:philiptaron/llama.cpp/nix#cuda --print-out-paths | xargs nix path-info --human-readable --closure-size 
/nix/store/7m5wi190al7m8ybd5i4hjf99w3nirkn8-llama.cpp      2.0G
```

SomeoneSerge committed:
flake.nix: avoid re-evaluating nixpkgs too many times

SomeoneSerge committed:
flake.nix: use flake-parts

SomeoneSerge committed:
*.nix: nixfmt

nix shell github:piegamesde/nixfmt/rfc101-style --command \
    nixfmt flake.nix .devops/nix/*.nix

SomeoneSerge committed:
nix: migrate to pname+version

SomeoneSerge committed:
flake.nix: overlay: expose both the namespace and the default attribute

SomeoneSerge committed:
ci: add the (Nix) flakestry workflow

SomeoneSerge committed:
flake.nix: add maintainers

SomeoneSerge committed:
nix: move meta down to follow Nixpkgs style more closely

SomeoneSerge committed:
nix: add missing meta attributes

nix: clarify the interpretation of meta.maintainers

nix: clarify the meaning of "broken" and "badPlatforms"

nix: passthru: expose the use* flags for inspection

E.g.:

```
❯ nix eval .#cuda.useCuda
true
```

SomeoneSerge commented:
The cuda build breaks after rebasing on https://github.com/ggerganov/llama.cpp/pull/4606/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR305-R306 , tracking in https://gitlab.kitware.com/cmake/cmake/-/issues/25536

SomeoneSerge committed:
flake.lock: update to hotfix CUDA::cuda_driver

Required to support https://github.com/ggerganov/llama.cpp/pull/4606

SomeoneSerge committed:
nix: cmakeFlags: explicit OFF bools

SomeoneSerge committed:
nix: cuda: reduce runtime closure

SomeoneSerge committed:
nix: fewer rebuilds

SomeoneSerge commented:
Rebased on master, bumped the `nixpkgs` input to include a workaround for the `CUDA::cuda_driver` issue.

Closed https://github.com/ggerganov/llama.cpp/pull/4632 and https://github.com/ggerganov/llama.cpp/pull/4633 because we've incorporated those changes into philiptaron's branch. The CI for the flake will come in a separate PR.

Confirming that things work (and `llama-cpp-cuda` is another 50% lighter):

```bash
❯ nix build .# .#cuda .#rocm --print-out-paths -L | xargs nix path-info --closure-size --human-readable 
/nix/store/1b6qcr7l9p0yx4mji8c6pbjcybz4fp88-llama-cpp-rocm-0.0.0          10.6G
/nix/store/61xjz6l35r1c6xp4v5py8iaia18bs5xy-llama-cpp-cuda-0.0.0         985.6M
/nix/store/ndx1xc79in3kmf65r4h1a28j3i72yv8q-llama-cpp-blas-0.0.0         340.1M
```

SomeoneSerge committed:
nix: respect config.cudaCapabilities

Green-Sky commented:
running llama cuda with this flake results in runtime errors:
```
CUDA error: cudaGetDevice(&current_device): CUDA driver is a stub library
  in function ggml_cuda_set_device at /build/source/ggml-cuda.cu:519
GGML_ASSERT: /build/source/ggml-cuda.cu:239: !"CUDA error"
```

SomeoneSerge commented:
Thanks @Green-Sky, we should strip the stub libraries' RUNPATHs from the binaries. They normally don't end up there because it's rare to link libcuda.so directly

SomeoneSerge committed:
nix: add the impure driver's location to the DT_RUNPATHs

Green-Sky commented:
@SomeoneSerge yes, works now.

Green-Sky reviewed (approved):
LGTM from my side, I tested default and cuda on my nixos.

philiptaron commented:
I also just tested all of this with aarch64-darwin (Mac M2 Air.) 

```
nix run github:philiptaron/llama.cpp/nix -- --file prompt.txt --model Downloads/daringmaid-13b.Q5_K_S.gguf --instruct --n-predict 2000 --color
```

SomeoneSerge committed:
nix: explicit jetson support

SomeoneSerge committed:
nix: clean sources more thoroughly

...this way outPaths change less frequently,
and so there are fewer rebuilds

SomeoneSerge committed:
nix: explicit mpi support

SomeoneSerge committed:
flake.nix: darwin: only expose the default

ParetoOptimalDev commented:
The current main branch doesn't build for me with Nix anymore, so I tried this one.

This one works great using:

```
NIXPKGS_ALLOW_UNFREE=1 nix run --impure github:ggerganov/llama.cpp/pulls/4605/head#cuda 
```

philiptaron commented:
@ggerganov I think this is good for merge; @SomeoneSerge and I are "pencils down". I've updated the PR description to encompass what we did over the past week.

ParetoOptimalDev commented:
> It is not cached (non-free) + it takes 20min to download and install everytime + its not very small and takes up alot of space.

@Green-Sky Not sure if anyone already mentioned this, but try adding these substituters for a high chance of getting a cache hit on cuda:

```
        "https://ai.cachix.org"
        "https://cuda-maintainers.cachix.org"
        "https://nix-community.cachix.org"
```

and keys:

```
        "cuda-maintainers.cachix.org-1:0dq3bujKpuEPMCX6U4WylrUDZ9JyUG0VpVZa7CNfq5E="
        "nix-community.cachix.org-1:mB9FSh9qf2dCimDSUo8Zy7bkq5CX+/rkCWyvRCYg3Fs="
        "ai.cachix.org-1:N9dzRK+alWwoKXQlnn0H6aUx0lU/mspIoz8hMvGvbbc="
```
